tests/integration/plugins/helpers.py

```diff
@@ -0,0 +1,188 @@
+from tests.integration.helpers import http_request
+
+logger = logging.getLogger(__name__)
+
+
+@retry(
+    wait=wait_fixed(wait=5) + wait_random(0, 5),
+    stop=stop_after_attempt(15),
+)
+async def wait_all_units_restarted_or_fail(active_since: Dict[Any, str]) -> Dict[Any, str]:
+    """Waits for all the units of the application to restart or fails.
+
+    Args:
+        active_since: dict, dictionary containing each unit and its latest timestamp as string.
+                            if unit's value is None, then set found timestamp.
+
+    Raises:
+        Exception: raised if action fails or if timestamp shows restart did not happen yet
+                   https://github.com/juju/python-libjuju/blob/ \
+                       48570bb8d51d38c430d12b86e39706ffd6969fcc/juju/unit.py#L301
+    """
+    result = {}
+    for unit, timestamp in active_since.items():
+        output = await unit.run(
+            "systemctl show snap.opensearch.daemon --property=ActiveEnterTimestamp"
+        )
+        timestamp = output.results["Stdout"].split("ActiveEnterTimestamp=")[1].rstrip()
+        if not active_since[unit] or active_since[unit] < timestamp:
+            result[unit] = timestamp
+        else:
+            raise Exception()
+    return result
+
+
+def generate_bulk_training_data(
+    n: int, ndims: int, index_name: str, vector_name: str, has_result: bool = False
+) -> str:
+    random.seed("seed")
+    print("The seed for randomness is: 'seed'")
+
+    data = random.randbytes(n * ndims)
+    if has_result:
+        responses = random.randbytes(n)
+    result = ""
+    result_list = []
+    for i in range(n):
+        result += json.dumps({"index": {"_index": index_name, "_id": i}}) + "\n"
+        result_list.append([float(data[j]) for j in range(i * ndims, (i + 1) * ndims)])
+        inter = {vector_name: result_list[i]}
+        if has_result:
+            inter["price"] = float(responses[i])
+        result += json.dumps(inter) + "\n"
+    return result, result_list
+
+
+@retry(
+    wait=wait_fixed(wait=5) + wait_random(0, 5),
+    stop=stop_after_attempt(15),
+)
+async def run_knn_training(
+    ops_test: OpsTest,
+    app: str,
+    unit_ip: str,
+    model_name: str,
+    payload: Dict[str, Any],
+) -> Optional[List[Dict[str, Any]]]:
+    """Sets models."""
+    endpoint = f"https://{unit_ip}:9200/_plugins/_knn/models/{model_name}/_train"
+
+    resp = await http_request(ops_test, "POST", endpoint, payload=payload, app=app)
+    return resp
+
+
+@retry(
+    wait=wait_fixed(wait=5) + wait_random(0, 5),
+    stop=stop_after_attempt(15),
+)
+async def wait_for_knn_training(
+    ops_test: OpsTest,
+    app: str,
+    unit_ip: str,
+    model_name: str,
+) -> bool:
+    """Train models."""
+    endpoint = f"https://{unit_ip}:9200/_plugins/_knn/models/{model_name}"
+
+    resp = await http_request(ops_test, "GET", endpoint, app=app)
+    return "created" in resp.get("state", "")
```

Comment:
could you follow the same pattern as [here](https://github.com/canonical/opensearch-operator/blob/main/tests/integration/ha/helpers.py#L343) ? this way we can remove the subsequent sleep used in the test after calling this (when `created not in resp.get(...` raise an exception so that it gets retried

Reply:
I removed the sleep and the for loops as either ```wait_for_knn_training``` throws an exception caught by retry (or eventually fails the test altogether); or responds with a valid HTTP message, but not containing the entry ```ctate: created```